### DISCUSSION TITLE
Metal 3.1

### DISCUSSION THREAD

**Discussion by rcaridade145** [KhronosGroup/MoltenVK#1938] [General]
Metal Shading Language seems to have been updated to 3.1 - https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf

Metal 3.1 introduces new atomic texture member functions supported on 1D texture, 1D texture
array, 2D texture, 2D texture array, 3D texture, and texture buffer for int, uint, ulong color
types:
• atomic_load - atomic load from a texture,
• atomic_store – atomic store to a texture,
• atomic_exchange - atomic exchange a value for a texture,
• atomic_compare_exchange_weak - atomic compare and exchange in a texture,
• atomic_fetch_op_explicit – atomic fetch and modify where op can be add, and,
max, min, or, sub, or xor for int and uint color type.
• atomic_max - atomic max in a texture for ulong color type.
• atomic_min - atomic min in a texture for ulong color type.

Under 6.16.1 Encoding Render Commands in Indirect Command Buffers

// Metal 3.1: Supported passing vertex strides
template <typename T ...>
void set_vertex_buffer(device T *buffer, size_t stride,
uint index);
template <typename T ...>
void set_vertex_buffer(constant T *buffer, size_t stride,
uint index);

Bing summarizes it with :

Thank you for providing the link. According to the current web page context, the PDF you provided is the Metal Shading Language Specification. It is a document that describes the Metal shading language and its features. The latest version of Metal shading language is 3.1 and it introduces new features such as:

Support for new data types such as 16-bit floating-point numbers and 64-bit integers.
New built-in functions for working with 16-bit floating-point numbers.
New built-in functions for working with 64-bit integers.
New built-in functions for working with matrices.
New built-in functions for working with quaternions.

**Comment by spnda** [comment]
The summary from Bing is completely wrong. People should really stop blindly using GPT. Here's everything what's new with Metal 3.1:
- 32-bit and 64-bit atomic ops on textures (1D, 2D, 3D) and texture arrays (1D, 2D)
- `ulong4` for sparse textures
- 16-bit "brain float" (`bfloatn`)
- Ray intersections with curves
- Built-in variables can now be declared globally except for `amplification_count`, `amplification_id`, and `payload`. (Could drastically simplify SPIRV-Cross when functions use `gl_InvocationIndex`, ...)
- Dynamic vertex buffer stride
- `CAMetalDisplayLink` used for frame pacing

**Reply by billhollings** [reply]
Thanks for the summary. Other than the MSL 3.1 doc above, where did you source this info?

**Reply by spnda** [reply]
I used the "Show Changes" feature on the Apple documentation site: https://developer.apple.com/documentation/metal?changes=latest_minor
You can essentially view a diff between different Xcode versions for any Framework.

Some more things:
- MTLArchitecture. Only has a name field.
- RG11B10 and RGB9E5 vertex formats (#1940)
- Indirect motion acceleration structure builds

**Reply by rcaridade145** [reply]
Thank you for this. Wasn't aware of that feature.

**Reply by billhollings** [reply]
Well! I learned something today! Thanks for pointing that documentation feature out!

**Comment by UltraHDR** [comment]
Not exactly Metal 3.1 specific, but since macOS 13.3 there's "shouldMaximizeConcurrentCompilation" https://developer.apple.com/documentation/metal/mtldevice/4133938-shouldmaximizeconcurrentcompilat which could reduce shader stutters. It was highlighed in https://developer.apple.com/videos/play/wwdc2023/10127/?time=540
